File: app/src/main/java/com/jmalinen/blackjack/model/Shoe.kt
package com.jmalinen.blackjack.model

class Shoe(private val numberOfDecks: Int) {
    private val cards: MutableList<Card> = mutableListOf()
    private var cutCardPosition: Int = 0
    val totalCards: Int = numberOfDecks * 52
    val cardsRemaining: Int get() = cards.size
    val penetration: Float get() = 1f - (cards.size.toFloat() / totalCards)

    init {
        shuffle()
    }

    fun shuffle() {
        cards.clear()
        for (deck in 0 until numberOfDecks) {
            for (suit in Suit.entries) {
                for (rank in Rank.entries) {
                    cards.add(Card(rank, suit))
                }
            }
        }
        cards.shuffle()
        cutCardPosition = (totalCards * 0.25).toInt()
    }

    fun draw(): Card {
        check(cards.isNotEmpty()) { "Shoe is empty" }
        return cards.removeFirst()
    }

    fun drawMatching(predicate: (Card) -> Boolean): Card {
        val index = cards.indexOfFirst(predicate)
        check(index != -1) { "No matching card in shoe" }
        return cards.removeAt(index)
    }

    fun needsReshuffle(): Boolean = cards.size <= cutCardPosition
}

File: ios/Blackjack/Blackjack/Model/Shoe.swift
class Shoe {
    private var cards: [Card] = []
    private var cutCardPosition: Int = 0
    let totalCards: Int
    var cardsRemaining: Int { cards.count }
    var penetration: Float { 1.0 - (Float(cards.count) / Float(totalCards)) }

    private let numberOfDecks: Int

    init(numberOfDecks: Int) {
        self.numberOfDecks = numberOfDecks
        self.totalCards = numberOfDecks * 52
        shuffle()
    }

    func shuffle() {
        cards.removeAll()
        for _ in 0..<numberOfDecks {
            for suit in Suit.allCases {
                for rank in Rank.allCases {
                    cards.append(Card(rank: rank, suit: suit))
                }
            }
        }
        cards.shuffle()
        cutCardPosition = Int(Double(totalCards) * 0.25)
    }

    func draw() -> Card {
        precondition(!cards.isEmpty, "Shoe is empty")
        return cards.removeFirst()
    }

    func drawMatching(_ predicate: (Card) -> Bool) -> Card {
        guard let index = cards.firstIndex(where: predicate) else {
            preconditionFailure("No matching card in shoe")
        }
        let result = cards.remove(at: index)
        cards.shuffle()
        return result
    }

    func needsReshuffle() -> Bool {
        cards.count <= cutCardPosition
    }
}
